File: golang/day3/main.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
)

func main() {
	data := readFile("../../input/day3.txt")
	fmt.Println(Part1(data))
	fmt.Println(Part2(data))
}

func readFile(filename string) []string {
	file, _ := os.Open(filename)
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var lines []string
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	if err := scanner.Err(); err != nil {
		fmt.Println("Error reading file:", err)
	}
	return lines
}

func Part1(data []string) int {

	count := 0

	for _, line := range data {
		best, _ := strconv.Atoi(string(line[0]))
		largest := math.MinInt64

		for i := 1; i < len(line); i++ {
			current, _ := strconv.Atoi(string(line[i]))
			largest = int(math.Max(float64(largest), float64(best*10+current)))
			best = int(math.Max(float64(current), float64(best)))
		}

		count += largest

	}

	return count
}

func Part2(data []string) int {
	count := 0
	turns := 12

	for _, line := range data {
		best := make([]int, turns)
		for i := range best {
			best[i] = -1
		}

		for i := 0; i < len(line); i++ {
			current := int(line[i] - '0')
			for j := turns - 1; j >= 0; j-- {
				if j == 0 {
					best[0] = max(best[0], current)
				} else if best[j-1] != -1 {
					best[j] = max(best[j], best[j-1]*10+current)
				}
			}
		}

		count += best[turns-1]

	}
	return count
}

File: python/day3/main.py
from collections.abc import Generator


# read file with generator expression
def read_file(filename: str) -> Generator[str, None, None]:
    with open(filename, "r", encoding="utf-8") as file:
        for line in file:
            yield line.strip()


inputPath = "../../input/day3.txt"


def part1() -> int:
    joltages = read_file(inputPath)
    count: int = 0

    for joltage in joltages:
        # largest: int = 0

        # for i in range(len(joltage)):
        #     for j in range(i + 1, len(joltage)):
        #         largest = max(largest, int(joltage[i] + joltage[j]))

        # print(largest)
        # count += largest
        #
        best_tens = int(joltage[0])
        largest = 0

        for i in range(1, len(joltage)):
            current = int(joltage[i])
            largest = max(largest, best_tens * 10 + current)
            best_tens = max(best_tens, current)

        count += largest

    return count


def part2() -> int:
    joltages = read_file(inputPath)
    count: int = 0
    turns = 12

    for joltage in joltages:
        best = [-1] * turns

        for digit in joltage:
            current = int(digit)
            for i in range(turns - 1, -1, -1):
                if i == 0:
                    best[0] = max(best[0], current)
                elif best[i - 1] != -1:
                    best[i] = max(best[i], best[i - 1] * 10 + current)

        count += best[-1]
    return count


if __name__ == "__main__":
    print(part1())
    print(part2())
